File: legacy/src/main/java/io/github/moulberry/moulconfig/gui/GuiComponentWrapper.java
/*
 * Copyright (C) 2023 NotEnoughUpdates contributors
 *
 * This file is part of MoulConfig.
 *
 * MoulConfig is free software: you can redistribute it
 * and/or modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation, either
 * version 3 of the License, or (at your option) any later version.
 *
 * MoulConfig is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with MoulConfig. If not, see <https://www.gnu.org/licenses/>.
 *
 */

package io.github.moulberry.moulconfig.gui;

import io.github.moulberry.moulconfig.internal.ForgeRenderContext;
import lombok.NonNull;
import lombok.ToString;
import net.minecraft.client.gui.GuiScreen;
import org.jetbrains.annotations.NotNull;
import org.lwjgl.input.Keyboard;
import org.lwjgl.input.Mouse;

import java.io.IOException;

@ToString
public class GuiComponentWrapper extends GuiScreen {
    @NonNull
    public GuiContext context;

    public GuiComponentWrapper(@NotNull GuiContext context) {
        this.context = context;
        context.setCloseRequestHandler(this::requestClose);
    }

    protected GuiImmediateContext createContext() {
        int x = Mouse.getEventX() * this.width / this.mc.displayWidth;
        int y = this.height - Mouse.getEventY() * this.height / this.mc.displayHeight - 1;
        return new GuiImmediateContext(
            new ForgeRenderContext(), 0, 0,
            width, height,
            x, y, x, y
        );
    }

    @Override
    public void onGuiClosed() {
        super.onGuiClosed();
        context.onAfterClose();
    }

    public void requestClose() {
        if (context.onBeforeClose() == CloseEventListener.CloseAction.NO_OBJECTIONS_TO_CLOSE) {
            this.mc.displayGuiScreen(null);
            if (this.mc.currentScreen == null) {
                this.mc.setIngameFocus();
            }
        }
    }

    @Override
    protected void keyTyped(char typedChar, int keyCode) {
        if (keyCode == Keyboard.KEY_ESCAPE) {
            requestClose();
        }
    }

    @Override
    public void handleKeyboardInput() throws IOException {
        super.handleKeyboardInput();

        if (Keyboard.getEventKeyState())
            context.getRoot().keyboardEvent(new KeyboardEvent.CharTyped(Keyboard.getEventCharacter()), createContext());
        context.getRoot().keyboardEvent(new KeyboardEvent.KeyPressed(Keyboard.getEventKey(), Keyboard.getEventKeyState()), createContext());
    }

    @Override
    public void drawScreen(int mouseX, int mouseY, float partialTicks) {
        super.drawScreen(mouseX, mouseY, partialTicks);
        ForgeRenderContext frc = new ForgeRenderContext();
        context.getRoot().render(new GuiImmediateContext(
            frc,
            0, 0, width, height, mouseX, mouseY, mouseX, mouseY
        ));
        frc.doDrawTooltip();
    }

    @Override
    public void handleMouseInput() throws IOException {
        super.handleMouseInput();
        if (Mouse.getEventButton() != -1) {
            context.getRoot().mouseEvent(new MouseEvent.Click(Mouse.getEventButton(), Mouse.getEventButtonState()), createContext());
        }
        if (Mouse.getEventDWheel() != 0) {
            context.getRoot().mouseEvent(new MouseEvent.Scroll((float) Mouse.getEventDWheel()), createContext());
        }
        if (Mouse.getEventDX() != 0 || Mouse.getEventDY() != 0) {
            context.getRoot().mouseEvent(new MouseEvent.Move((float) Mouse.getEventDX(), (float) Mouse.getEventDY()), createContext());
        }
    }
}

File: modern/src/main/kotlin/io/github/notenoughupdates/moulconfig/gui/GuiComponentWrapper.kt
package io.github.notenoughupdates.moulconfig.gui

import io.github.moulberry.moulconfig.gui.GuiContext
import io.github.moulberry.moulconfig.gui.GuiImmediateContext
import io.github.moulberry.moulconfig.gui.KeyboardEvent
import io.github.moulberry.moulconfig.gui.MouseEvent
import io.github.notenoughupdates.moulconfig.platform.ModernRenderContext
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.gui.screen.Screen
import net.minecraft.text.Text

class GuiComponentWrapper(
    val context: GuiContext,
    label: Text = Text.literal("")
) : Screen(label) {
    private fun createContext(drawContext: DrawContext? = null): GuiImmediateContext {
        val mouse = MinecraftClient.getInstance().mouse
        val window = client!!.window
        val x = (mouse.x * window.scaledWidth.toDouble() / window.width.toDouble()).toInt()
        val y = (mouse.y * window.scaledHeight.toDouble() / window.height.toDouble()).toInt()
        return GuiImmediateContext(
            ModernRenderContext(
                drawContext ?: DrawContext(
                    MinecraftClient.getInstance(),
                    MinecraftClient.getInstance().bufferBuilders.entityVertexConsumers
                )
            ),
            0, 0,
            window.scaledWidth,
            window.scaledHeight,
            x, y, x, y
        )
    }

    override fun render(drawContext: DrawContext?, i: Int, j: Int, f: Float) {
        super.render(drawContext, i, j, f)
        val ctx = createContext(drawContext)
        context.root.render(ctx)
        ctx.renderContext.doDrawTooltip()
    }

    override fun charTyped(c: Char, i: Int): Boolean {
        context.root.keyboardEvent(KeyboardEvent.CharTyped(c), createContext())
        return true
    }

    override fun keyPressed(i: Int, j: Int, k: Int): Boolean {
        context.root.keyboardEvent(KeyboardEvent.KeyPressed(i, true), createContext())
        return true
    }

    override fun keyReleased(i: Int, j: Int, k: Int): Boolean {
        context.root.keyboardEvent(KeyboardEvent.KeyPressed(i, false), createContext())
        return true
    }

    override fun mouseMoved(d: Double, e: Double) {
        val ctx = createContext()
        val di = d.toInt()
        val ei = e.toInt()
        val event = MouseEvent.Move(d.toFloat() - ctx.mouseX, e.toFloat() - ctx.mouseY)
        val repositionedContext = ctx.copy(mouseX = di, mouseY = ei, absoluteMouseX = di, absoluteMouseY = ei)
        context.root.mouseEvent(event, repositionedContext)
    }

    override fun mouseClicked(d: Double, e: Double, i: Int): Boolean {
        context.root.mouseEvent(MouseEvent.Click(i, true), createContext())
        return true
    }

    override fun mouseReleased(d: Double, e: Double, i: Int): Boolean {
        context.root.mouseEvent(MouseEvent.Click(i, false), createContext())
        return true
    }

    override fun mouseScrolled(d: Double, e: Double, f: Double): Boolean {
        context.root.mouseEvent(MouseEvent.Scroll(f.toFloat()), createContext())
        return true
    }

    override fun mouseDragged(d: Double, e: Double, i: Int, f: Double, g: Double): Boolean {
        return true
    }
}
